skada/metrics.py

```diff
@@ -782,3 +784,131 @@ def _score(self, estimator, X, y=None, sample_domain=None, **params):
+        self.p = p
+        self.threshold = threshold
+        self._sign = 1 if greater_is_better else -1
+
+        if self.p <= 0:
+            raise ValueError("The order of the p-norm must be positive")
+
+    def _score(self, estimator, X, y, sample_domain=None, **params):
+        if not hasattr(estimator, "predict_proba"):
+            raise AttributeError(
+                "The estimator passed should have a 'predict_proba' method. "
+                f"The estimator {estimator!r} does not."
+            )
+
+        X, y, sample_domain = check_X_y_domain(X, y, sample_domain, allow_nd=True)
+        source_idx = extract_source_indices(sample_domain)
+
+        # Check from y values if it is a classification problem
+        y_type = _find_y_type(y)
+        if y_type != Y_Type.DISCRETE:
+            raise ValueError("MaNo scorer only supports classification problems.")
+
+        # Recover target probabilities
+        if not isinstance(estimator, Pipeline):
+            # 1) Recover logits on target
+            logits = estimator.infer(X[~source_idx], **params).cpu().detach().numpy()
+
+            # 2) Normalize logits to obtain probabilities
+            criterion = self._get_criterion(logits)
+            proba = self._softrun(
+                logits=logits,
+                criterion=criterion,
+                threshold=self.threshold,
+            )
+        else:
+            # Directly recover predicted probabilities
+            proba = estimator.predict_proba(
+                X[~source_idx], sample_domain=sample_domain[~source_idx], **params
+            )
+
+        # 3) Aggregate following Eq.(2) of [37]_.
+        score = np.mean(proba**self.p) ** (1 / self.p)
+
+        return self._sign * score
+
+    def _get_criterion(self, logits):
+        """
+        Compute criterion to select the proper normalization.
+        See Eq.(6) of [1]_ for more details.
+        """
+        proba = self._stable_softmax(logits)
+        proba = np.log(proba)
+        divergence = -np.mean(proba)
+
+        return divergence
+
+    def _softrun(self, logits, criterion, threshold):
+        """Normalize the logits following Eq.(6) of [37]_."""
+        if criterion > threshold:
+            # Apply softmax normalization
+            outputs = self._stable_softmax(logits)
+
+        else:
+            # Apply Taylor approximation
+            outputs = self._taylor_softmax(logits)
+
+        return outputs
+
+    @staticmethod
+    def _stable_softmax(logits):
+        """Compute softmax function."""
+        logits -= np.max(logits, axis=1, keepdims=True)
+        exp_logits = np.exp(logits)
+        exp_logits /= np.sum(exp_logits, axis=1, keepdims=True)
+        return exp_logits
+
+    @staticmethod
+    def _taylor_softmax(logits):
+        """Compute Taylor approximation of order 2 of softmax."""
+        tay_logits = 1 + logits + logits**2 / 2
+        tay_logits -= np.min(tay_logits, axis=1, keepdims=True)
```

Comment:
Where does the min come from in the eq ? I didn't find it in the paper

Reply:
It is not needed when p=2 but necessary for p>=3 to ensure that the outputs is a probability. It is explained in footnote 2 of page 6 in https://arxiv.org/pdf/2405.18979